src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -233,12 +234,9 @@ public void stop (int delay) {
         terminating = true;
         try { schan.close(); } catch (IOException e) {}
         selector.wakeup();
-        long latest = System.currentTimeMillis() + delay * 1000;
-        while (System.currentTimeMillis() < latest) {
+        long latest = System.nanoTime() + Duration.ofSeconds(delay).toNanos();
+        while (activeExchanges() > 0 && System.nanoTime() < latest) {
```

Comment:
I don't think that works unfortunately. We need something that takes into account that an exchange may have been started but hasn't reached the point where startExchange has been called. 

The problem is that exchangeCount is incremeted asynchronously in the ExchangeImpl constructor, and the ExchangeImpl is created by the Exchange::run method - which is run asynchronoously in the executor (submitted by the dispacher thread).

Possibly posting a StopRequested event to the dispatcher thread, and have the dispacther thread switch finished = true when is sees the StopRequested event and notices that allConnections only contains idleConnections (in which case we could also assert that exchangeCount == 0).

Unless I'm not mistaken this is not going to be a trivial fix.

Reply:
> Unless I'm not mistaken this is not going to be a trivial fix.

Alright, I'll convert this to a draft PR for now. I may explore solutions, but it seems a full fix could be above my pay grade.